pandas/_libs/tslibs/_timestamp.pyx

```diff
@@ -0,0 +1,420 @@
+            other_int = other.astype('timedelta64[ns]').view('i8')
+            return self.__class__(self.value + other_int,
+                                  tz=self.tzinfo, freq=self.freq)
+
+        elif is_integer_object(other):
+            maybe_integer_op_deprecated(self)
+
+            if self is NaT:
+                # to be compat with Period
+                return NaT
+            elif self.freq is None:
+                raise ValueError("Cannot add integral value to Timestamp "
+                                 "without freq.")
+            return self.__class__((self.freq * other).apply(self),
+                                  freq=self.freq)
+
+        elif PyDelta_Check(other) or hasattr(other, 'delta'):
+            # delta --> offsets.Tick
+            # logic copied from delta_to_nanoseconds to prevent circular import
+            if hasattr(other, 'nanos'):
+                nanos = other.nanos
+            elif hasattr(other, 'delta'):
+                nanos = other.delta
+            elif PyDelta_Check(other):
+                nanos = (other.days * 24 * 60 * 60 * 1000000 +
+                         other.seconds * 1000000 +
+                         other.microseconds) * 1000
+
+            result = self.__class__(self.value + nanos,
+                                    tz=self.tzinfo, freq=self.freq)
+            if getattr(other, 'normalize', False):
+                # DateOffset
+                result = result.normalize()
+            return result
+
+        # index/series like
+        elif hasattr(other, '_typ'):
+            return NotImplemented
+
+        result = datetime.__add__(self, other)
+        if PyDateTime_Check(result):
+            result = self.__class__(result)
+            result.nanosecond = self.nanosecond
+        return result
+
+    def __sub__(self, other):
+        if (is_timedelta64_object(other) or is_integer_object(other) or
+                PyDelta_Check(other) or hasattr(other, 'delta')):
+            # `delta` attribute is for offsets.Tick or offsets.Week obj
+            neg_other = -other
+            return self + neg_other
+
+        typ = getattr(other, '_typ', None)
+
+        # a Timestamp-DatetimeIndex -> yields a negative TimedeltaIndex
+        if typ in ('datetimeindex', 'datetimearray'):
+            # timezone comparison is performed in DatetimeIndex._sub_datelike
+            return -other.__sub__(self)
+
+        # a Timestamp-TimedeltaIndex -> yields a negative TimedeltaIndex
+        elif typ in ('timedeltaindex', 'timedeltaarray'):
+            return (-other).__add__(self)
+
+        elif other is NaT:
+            return NaT
+
+        # coerce if necessary if we are a Timestamp-like
+        if (PyDateTime_Check(self)
+                and (PyDateTime_Check(other) or is_datetime64_object(other))):
+            if isinstance(self, _Timestamp):
+                other = self.__class__(other)
+            else:
+                self = other.__class__(self)
```

Comment:
```
self = Timestamp(self)
other = Timestamp(other)
```
->
```
if isinstance(self, _Timestamp):
    other = self.__class__(other)
 else:
     self = other.__class__(self)
```
Both can't blindly be converted to `Timestamp` because `__class__` is being used.

Reply:
are there cases where `isinstance(self, _Timestamp)` is False?  If so, can you add a comment above that check referencing the relevant cython semantics.  If not, can you remove the check.

Reply:
Yes, for instance without the `else` the `TestTimedelta64ArithmeticUnsorted.test_subtraction_ops_with_tz` test fails